Dhis2

kaapstorm opened the pull request:
Fixed the way forms are exported to DHIS2. Moved settings to fixtures so that web users can set them.

kaapstorm committed:
DHIS2: *WIP* Changes for forms

Risk Assessment program events come from risk assessment forms. Need
better way of iterating forms, and storing which forms have been
processed.

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: WIP: Tasks updated, (unnecessary) views created, chunks commented out

kaapstorm committed:
Add HTTP Basic Auth to form and case repeaters

kaapstorm committed:
DHIS2: Use payload generators to create DHIS2 program events

kaapstorm committed:
DHIS2: Fixed adding tracked entity instances

kaapstorm committed:
DHIS2: Fixed "CCHQ Case ID" spelling, corrected return value of update_te_inst

kaapstorm committed:
DHIS2: Use fallback org unit if case missing one

kaapstorm committed:
DHIS2: Enrolling, pulling, pushing fixed

kaapstorm committed:
DHIS2: Use dhis_org_id custom user field

kaapstorm committed:
DHIS2: Syncing forms almost working. WIP

kaapstorm committed:
DHIS2: JsonFormRepeater added

kaapstorm committed:
DHIS2: check cases every 5 mins, to keep in time with forwarded forms

kaapstorm committed:
DHIS2: Use fixtures for settings

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: Fix missing brackets. Default to test domain.

kaapstorm committed:
DHIS2: Moved JsonFormRepeater class, fixed tests

kaapstorm committed:
DHIS2: Fixed JsonFormRepeater bug

dimagimon commented:
There are 5 errors:

* custom/dhis2/__init__.py, line 4 - F401 'payload_generators' imported but unused
* custom/dhis2/const.py, line 71 - W391 blank line at end of file
* custom/dhis2/tests.py, line 10 - F401 'RepeatRecord' imported but unused
* custom/dhis2/tests.py, line 11 - F401 'create_repeat_records' imported but unused
* custom/dhis2/tests.py, line 15 - F401 'JsonFormRepeater' imported but unused

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
Merge branch 'master' into dhis2

snopoke commented:
@kaapstorm is this GTG or are there still outstanding changes?

kaapstorm commented:
This is GTG. There is a change I want to make, to move settings behind a feature flag, but I need to balance that with the priorities of the next two things on my TODO list.

kaapstorm committed:
Merge branch 'master' into dhis2

snopoke commented on corehq/apps/receiverwrapper/models.py:
```diff
@@ -337,6 +353,27 @@ def payload_doc(self, repeat_record):
 
 
+@register_repeater_type
+class JsonFormRepeater(FormRepeater):
```
What's the purpose of the `JsonFormRepeater`, it doesn't seem to do anything that the normal `FormRepeater` does. Can't you just register the `FormRepeaterDhis2EventPayloadGenerator` to the FormRepeater class?

kaapstorm replied:
It sets the content type to "application/json". Without that DHIS2 assumes XML, and throws an error when it can't parse it. It also prevents "app_id" from being appended to the URL in API calls.

snopoke replied:
Seems like that could be incorporated into the payload generator and an additional option to optionally include the app_id.

snopoke committed:
allow payload generators to contribute headers

snopoke committed:
make app_id url param optional

snopoke committed:
remove JsonFormRepeater class

snopoke committed:
fix circular import

dimagimon commented on custom/dhis2/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# Import payload_generators to register form repeater payload generators, so
+# that DHIS2 Nutrition Assessment and Risk Assessment events can be created
+# from CCHQ forms.
+import payload_generators
```
F401 'payload_generators' imported but unused

dimagimon commented on custom/dhis2/const.py:
```diff
@@ -0,0 +1,71 @@
+    # 'mother_first_name': 'Name of the Mother/Guardian',
+    # 'gn': 'GN Division of Household',
+}
+
```
W391 blank line at end of file

dimagimon commented on custom/dhis2/tests.py:
```diff
@@ -5,13 +5,14 @@
-from unittest import skip, SkipTest
+from unittest import skip
 from corehq.apps.fixtures.models import FixtureDataType, FixtureTypeField
+from corehq.apps.receiverwrapper.models import RepeatRecord
```
F401 'RepeatRecord' imported but unused

dimagimon commented on custom/dhis2/tests.py:
```diff
@@ -5,13 +5,14 @@
+from unittest import skip
 from corehq.apps.fixtures.models import FixtureDataType, FixtureTypeField
+from corehq.apps.receiverwrapper.models import RepeatRecord
+from corehq.apps.receiverwrapper.signals import create_repeat_records
```
F401 'create_repeat_records' imported but unused

kaapstorm committed:
Merge branch 'dhis2' of github.com:dimagi/commcare-hq into dhis2

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: Created Feature Flag

Moved settings into their own model
Settings managed from Project Settings when flag enabled.

kaapstorm committed:
DHIS2: Store parent org unit IDs

Search up org unit hierarchy if a user is not found at the current org unit.

dimagimon commented:
There are 1 errors:

* custom/dhis2/forms.py, line 30 - F841 local variable 'dhis2_data' is assigned to but never used

kaapstorm committed:
DHIS2: Removed unused var

kaapstorm committed:
DHIS2: Updated tests

kaapstorm committed:
Merge branch 'master' into dhis2

Conflicts:
	corehq/toggles.py

kaapstorm committed:
DHIS2: Replaced Dhis2SettingsSchema with DictProperty

Populate initial values from DB

dimagimon commented on corehq/apps/domain/views.py:
```diff
@@ -4,9 +4,12 @@
 import dateutil
 from django.core.mail import send_mail
 from django.core.paginator import Paginator
+from django.forms.forms import get_declared_fields
```
F401 'get_declared_fields' imported but unused

kaapstorm committed:
Merge branch 'master' into dhis2

snopoke commented on custom/dhis2/models.py:
```diff
@@ -1,13 +1,35 @@
-LOOKUP_TABLE = 'dhis2_org_unit'
+class Dhis2Settings(Document):
+    domain = StringProperty()
+    dhis2 = DictProperty()
```
Any reason to make this a dict vs individual properties or a sub-object?

snopoke commented:
Kicking build

czue commented on custom/dhis2/tasks.py:
```diff
@@ -288,64 +214,60 @@ def gen_children_only_ours(domain):
-def sync_child_entities():
+# TODO: Use case forwarding, or form forwarding of registration forms
+@periodic_task(run_every=timedelta(minutes=5))  # Run every 5 minutes to match forwarded forms
+def sync_cases():
```
@kaapstorm this is clogging up the celery queues and making other parts of the site not work. I just disabled it in https://github.com/dimagi/commcare-hq/pull/5550. I suspect the slowness is at least partly to do with the fact that you are looping through every domain in the database every 5 minutes (I didn't look closely enough to see whether it's also the fact that it's doing lots of work for the dhis domains or whether this issue caused additional slowness https://github.com/dimagi/commcare-hq/pull/5551). is the TODO above something we can change soon? otherwise we may need to either drastically reduce the time this function takes by optimizing it better or reduce the frequency significantly.

kaapstorm replied:
Thanks @czue. I'll get onto that now. There are two ways to improve performance here; using case/form forwarding is one, and using a single query to get the enabled domains is another. The query will be an easy win. I'll do that first.

snopoke replied:
How are you going to query to get all the enabled domains? As far as I can
see there isn't any view or index that would give you that information.
Actually I guess you can query the toggle to see which domains are enabled
in the toggle.

czue replied:
if the DHIS configs live in separate documents one simple thing would be to just move them to the meta database and make a view.
